### PR TITLE
Fix visualizations sometimes not opening with space bar.

### LINKED ISSUE #5223: Opening visualization with space does not always work.
This task is automatically imported from the old Task Issue Board and it was originally created by Adam Obuchowicz.
Original issue is [here](https://www.pivotaltracker.com/story/show/184266469).
<hr />

https://discord.com/channels/401396655599124480/1064967872012615872

There is no reliable reproduction scenario, but here are some clues:
* Sylwia reported, that she usually spots the issue when recording screens
* I (Adam) reproduced the issue in browser when build with `ide watch --skip-wasm-opt`.

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description

Visualizations closing right after opening was caused by the GUI being unresponsive during loading of some visualizations. This caused the timer for measuring the time between space bar press and space bar release to be inflated. The delayed events triggered the "visualization preview mode”, thus closing the visualization has it seemed that the space bar was held down, even though the events just arrived with some delay. 

The problem is mitigated by considering the number of frames that have passed between the space and down and the space bar up event, instead of just the wall clock time. If the number of frames is too low, this indicates that frames were dropped to the time is inflated. 

Fixes https://github.com/enso-org/enso/issues/5223

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: lib/rust/web/src/lib.rs [hunk 1]
```diff
@@ -947,3 +948,52 @@ impl TimeProvider for Performance {
...
+    _closure_handle:       Rc<RefCell<Option<Closure<(dyn FnMut(f64))>>>>,
+}
+
+impl FrameCounter {
+    /// Creates a new frame counter.
+    pub fn start_counting() -> Self {
+        let frames: Counter = default();
+        let frames_handle = Rc::clone(&frames);
+        let closure_handle = Rc::new(RefCell::new(None));
+        let closure_handle_internal = Rc::clone(&closure_handle);
```
Comment (kazcw) [inline lib/rust/web/src/lib.rs]: Is there a memory leak here? The closure strongly-references the closure handle, which owns the closure.

File: lib/rust/web/src/lib.rs [hunk 2]
```diff
@@ -947,3 +948,52 @@ impl TimeProvider for Performance {
...
+
+
+// ====================
+// === FrameCounter ===
+// ====================
+
+type Counter = Rc<std::cell::Cell<i32>>;
+
+#[derive(Debug)]
+/// A counter that counts frames thr frames that have passed since its initialization.
```
Comment (kazcw) [inline lib/rust/web/src/lib.rs]: Typo

**Review by kazcw — CHANGES_REQUESTED** [review]

File: app/gui/view/graph-editor/src/lib.rs [hunk 3]
```diff
@@ -94,6 +94,9 @@ pub mod prelude {
 
 const SNAP_DISTANCE_THRESHOLD: f32 = 10.0;
 const VIZ_PREVIEW_MODE_TOGGLE_TIME_MS: f32 = 300.0;
+/// Number of frames we expect to pass during the
+/// `VIZ_PREVIEW_MODE_TOGGLE_TIME_MS` interval. Assumes 60fps.  
+const VIZ_PREVIEW_MODE_TOGGLE_FRAMES: i32 = (VIZ_PREVIEW_MODE_TOGGLE_TIME_MS / 16.0) as i32;
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/lib.rs]: If we assume 60 FPS I would rather put here 1000.0/60.0. Or make assumed FPS another constant

**Review by farmaazon — APPROVED** [review]
Please address the leak issue by @kazcw

File: app/gui/view/graph-editor/src/lib.rs [hunk 4]
```diff
@@ -94,6 +94,10 @@ pub mod prelude {
 
 const SNAP_DISTANCE_THRESHOLD: f32 = 10.0;
 const VIZ_PREVIEW_MODE_TOGGLE_TIME_MS: f32 = 300.0;
+/// Number of frames we expect to pass during the
+/// `VIZ_PREVIEW_MODE_TOGGLE_TIME_MS` interval. Assumes 60fps.
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/lib.rs]: This lacks documentation of what these constants do. Please add docs to previous constant as well. Also, the explanation `Number of frames we expect to pass during the...` doesn't really explain what it means and what it is used for.

File: app/gui/view/graph-editor/src/lib.rs [hunk 5]
```diff
@@ -3446,10 +3450,25 @@ fn new_graph_editor(app: &Application) -> GraphEditor {
...
+            (time, Some(frame_counter))
+        });
     viz_release_time     <- viz_release . map(|_| web::window.performance_or_panic().now() as f32);
-    viz_press_time_diff  <- viz_release_time.map2(&viz_press_time,|t1,t0| t1-t0);
-    viz_preview_mode     <- viz_press_time_diff.map(|t| *t > VIZ_PREVIEW_MODE_TOGGLE_TIME_MS);
+    viz_preview_mode  <- viz_release_time.map2(&viz_press_time,|t1,(t0,counter)| {
+        let diff = t1-t0;
+        let long_enough = diff > VIZ_PREVIEW_MODE_TOGGLE_TIME_MS;
+        // We also check the number of passed frames to ensure we did not spend too much time in
+        // dropped frames during the visualisation initialisation.
```
Comment (wdanilo) [inline app/gui/view/graph-editor/src/lib.rs]: What does it mean? What is "too much time"? There is no explanation why this code exists and what it really does.

**Review by wdanilo — CHANGES_REQUESTED** [review]
I need a better explanation of why this PR works and what was the cause that visualizations were disappearing. The code is probably fine, but it should explain itself for future readers

**Comment by MichaelMauderer** [comment]
> I need a better explanation of why this PR works and what was the cause that visualizations were disappearing. The code is probably fine, but it should explain itself for future readers

The problem here is, that we are trying to measure the time between the key down and the key up event. If the key up event happens before the `VIZ_PREVIEW_MODE_TOGGLE_TIME_MS` has passed, we consider it a normal key press that opens the visualization. If the key up event happens after `VIZ_PREVIEW_MODE_TOGGLE_TIME_MS` has passed, we assume that a "press and hold” action was performed, which would trigger our visualization preview and thus the visualization should be closed when the key is release. But if the opening of the visualization takes a lot of time, specifically, longer than `VIZ_PREVIEW_MODE_TOGGLE_TIME_MS`, then the key up event will be processed too late, even if the key was already released right after the key down. We can guard against this by not just measuring the time, but checking that enough frames have passed. Since if we take too much time processing things, we get dropped frames. 

I'll now update the docs to make this clearer.

**Review by wdanilo — APPROVED** [review]
Perfect, now it is fully understandable, thanks!

**Comment by MichaelMauderer** [comment]
@kazcw Could you also re-review tour requested changes?

**Review by kazcw — APPROVED** [review]

**Review by vitvakatu — APPROVED** [review]
QA passed

**Pull request merged**